### PR TITLE
Allow creation of ADPiecewiseLinear

### LINKED ISSUE #20644: Allow creation of ADPiecewiseLinear functions
## Reason
Allows use of piecewise-linear functions where we want ADReal functors

## Design
Follow the design for ADParsedFunction

## Impact
Enhanced usability of functor system

### PR DESCRIPTION
**Pull request by lindsayad**
Refs joe61vette/ASP#7
Closes #20644

### CODE REVIEW

File: framework/src/functions/PiecewiseLinearBase.C [hunk 1]
```diff
@@ -43,51 +47,62 @@ PiecewiseLinearBase::buildInterpolation(const bool extrap)
...
+PiecewiseLinearBaseTempl<BaseClass>::value(Real t, const Point & p) const
 {
   const auto x = _has_axis ? p(_axis) : t;
   return _scale_factor * _linear_interp->sample(x);
 }
 
+template <typename BaseClass>
 ADReal
-PiecewiseLinearBase::value(const ADReal & t, const ADPoint & p) const
+PiecewiseLinearBaseTempl<BaseClass>::value(const ADReal & t, const ADPoint & p) const
```
Comment (GiudGiud) [inline framework/src/functions/PiecewiseLinearBase.C:60]: so the non AD version also has this member?

Comment (lindsayad) [reply framework/src/functions/PiecewiseLinearBase.C:60]: yes. This is to allow "hacks" for people like @joshuahansel who want to pass variables (evaluated at quadrature points) in as the `t` argument (and I think I've actually seen people pass vector-like variables in for `p` as well) and get AD operations

Comment (GiudGiud) [reply framework/src/functions/PiecewiseLinearBase.C:60]: can't @joshuahansel use ADFunctions for this though?
Are ADFunctions only a thing for our requirement to get a functor returning ADReal?

**Review by GiudGiud — APPROVED** [review]

File: framework/src/functions/PiecewiseLinearBase.C (hunk 1 above)
Comment (GiudGiud) [reply framework/src/functions/PiecewiseLinearBase.C:60]: let's resolve this for the merge and discuss later

**Pull request merged**